libdebug/interfaces/ptrace_interface.py

```diff
@@ -239,6 +239,55 @@ def step_until(self, thread: ThreadContext, address: int, max_steps: int):
         if result == -1:
             errno_val = self.ffi.errno
             raise OSError(errno_val, errno.errorcode[errno_val])
+        
+    def finish(self, thread: ThreadContext, exact: bool):
+        """Executes instructions of the specified thread until the current function returns.
+
+        Args:
+            thread (ThreadContext): The thread to step.
+            exact (bool): If True, the command is implemented as a series of `step` commands.
+        """
+        
+        if exact:
+
+            result = self.lib_trace.exact_finish(
+                self._global_state, thread.thread_id
+            )
+            
+            if result == -1:
+                errno_val = self.ffi.errno
+                raise OSError(errno_val, errno.errorcode[errno_val])
+        else:
+            # Breakpoint to return address
+            last_saved_instruction_pointer = thread.current_return_address()
+
+            # If a breakpoint already exists at the return address, we don't need to set a new one
+            found = False
+            ip_breakpoint = None
+            
+            for bp in self.context.breakpoints.values():
+                if bp.address == last_saved_instruction_pointer:
+                    found = True
+                    ip_breakpoint = bp
+                    break
+            
+            # Check if we have enough hardware breakpoints available
+            # Otherwise we use a software breakpoint
+            install_hw_bp = self.hardware_bp_helpers[thread.thread_id].available_breakpoints() > 0
```

Comment:
I would move this inside the if statement since it is only used when the breakpoint does not already exist. 

Additionally, I would consider the scenario where the bp does exist as a software breakpoint, and hardware breakpoints are available. In this case, it might be worth transforming the bp into a hardware breakpoint just for the finish.

Reply:
I don't know if it's a good idea to force to hardware a breakpoint that the user wanted as *software*. May cause confusion, as for moving inside the if, I'm on it.

Reply:
Makes sense too